File: shared/src/main/java/com/runtracker/shared/data/model/TrainingPlan.kt
package com.runtracker.shared.data.model

import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.TypeConverters
import com.runtracker.shared.data.db.Converters

@Entity(tableName = "training_plans")
@TypeConverters(Converters::class)
data class TrainingPlan(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val name: String,
    val description: String,
    val goalType: GoalType,
    val targetDistance: Double? = null,
    val targetTime: Long? = null,
    val startDate: Long,
    val endDate: Long,
    val weeklySchedule: List<ScheduledWorkout> = emptyList(),
    val isActive: Boolean = true,
    val createdAt: Long = System.currentTimeMillis()
)

enum class GoalType {
    FIRST_5K,
    IMPROVE_5K,
    FIRST_10K,
    IMPROVE_10K,
    HALF_MARATHON,
    MARATHON,
    GENERAL_FITNESS,
    WEIGHT_LOSS,
    CUSTOM
}

data class ScheduledWorkout(
    val id: String,
    val dayOfWeek: Int,
    val weekNumber: Int,
    val workoutType: WorkoutType,
    val targetDistanceMeters: Double? = null,
    val targetDurationMinutes: Int? = null,
    val targetPaceSecondsPerKm: Double? = null,
    val targetPaceMinSecondsPerKm: Double? = null,
    val targetPaceMaxSecondsPerKm: Double? = null,
    val targetHeartRateZone: HeartRateZone? = null,
    val targetHeartRateMin: Int? = null,
    val targetHeartRateMax: Int? = null,
    val intervals: List<Interval>? = null,
    val description: String,
    val isCompleted: Boolean = false,
    val completedRunId: Long? = null
)

enum class WorkoutType {
    // Basic runs
    EASY_RUN,
    LONG_RUN,
    RECOVERY_RUN,
    
    // Speed work
    TEMPO_RUN,
    INTERVAL_TRAINING,
    FARTLEK,
    RACE_PACE,
    
    // Strength & hills
    HILL_REPEATS,
    HILL_SPRINTS,
    STAIR_WORKOUT,
    
    // Specialized
    PROGRESSION_RUN,      // Start easy, finish fast
    NEGATIVE_SPLIT,       // Second half faster than first
    THRESHOLD_RUN,        // Sustained effort at lactate threshold
    VO2_MAX_INTERVALS,    // Short, very hard intervals
    YASSO_800S,           // Marathon predictor workout
    MILE_REPEATS,         // Classic speed workout
    LADDER_WORKOUT,       // Increasing/decreasing intervals
    PYRAMID_WORKOUT,      // Build up then back down
    
    // Race simulation
    RACE_SIMULATION,      // Practice race conditions
    TIME_TRIAL,           // All-out effort for distance
    PARKRUN_PREP,         // 5K race preparation
    
    // Recovery & maintenance
    SHAKE_OUT_RUN,        // Short, easy pre/post race
    BASE_BUILDING,        // Aerobic foundation
    AEROBIC_MAINTENANCE,  // Maintain fitness
    
    // Cross-training
    CROSS_TRAINING,
    CYCLING,
    SWIMMING,
    STRENGTH_TRAINING,
    YOGA_STRETCHING,
    
    // Rest
    REST_DAY,
    ACTIVE_RECOVERY,      // Very light activity
    
    // Custom
    CUSTOM                // User-designed workout
}

data class Interval(
    val type: IntervalType,
    val durationSeconds: Int? = null,
    val distanceMeters: Double? = null,
    val targetPaceSecondsPerKm: Double? = null,
    val targetPaceMinSecondsPerKm: Double? = null,
    val targetPaceMaxSecondsPerKm: Double? = null,
    val targetHeartRateZone: HeartRateZone? = null,
    val targetHeartRateMin: Int? = null,
    val targetHeartRateMax: Int? = null,
    val repetitions: Int = 1
)

enum class IntervalType {
    WARMUP,
    WORK,
    RECOVERY,
    COOLDOWN
}

enum class HeartRateZone(val displayName: String, val minPercent: Int, val maxPercent: Int) {
    ZONE_1("Recovery", 50, 60),       // 50-60% max HR - Recovery
    ZONE_2("Aerobic Base", 60, 70),   // 60-70% max HR - Aerobic base
    ZONE_3("Aerobic Capacity", 70, 80), // 70-80% max HR - Aerobic capacity
    ZONE_4("Threshold", 80, 90),      // 80-90% max HR - Threshold
    ZONE_5("VO2 Max", 90, 100)        // 90-100% max HR - VO2 Max
}

data class HeartRateZoneTime(
    val zone: HeartRateZone,
    val durationMillis: Long,
    val percentOfTotal: Double = 0.0
) {
    val durationFormatted: String get() {
        val totalSeconds = durationMillis / 1000
        val minutes = totalSeconds / 60
        val seconds = totalSeconds % 60
        return String.format("%d:%02d", minutes, seconds)
    }
}

File: iosApp/GoSteady/Sources/Models/TrainingPlan.swift
import Foundation
import SwiftData

@Model
final class TrainingPlan {
    @Attribute(.unique) var id: UUID
    var name: String
    var planDescription: String
    var goalType: GoalType
    var targetDistance: Double?
    var targetTime: Int64?
    var startDate: Date
    var endDate: Date
    var weeklySchedule: [ScheduledWorkout]
    var isActive: Bool
    var createdAt: Date

    init(
        id: UUID = UUID(),
        name: String,
        planDescription: String,
        goalType: GoalType,
        targetDistance: Double? = nil,
        targetTime: Int64? = nil,
        startDate: Date,
        endDate: Date,
        weeklySchedule: [ScheduledWorkout] = [],
        isActive: Bool = true,
        createdAt: Date = Date()
    ) {
        self.id = id
        self.name = name
        self.planDescription = planDescription
        self.goalType = goalType
        self.targetDistance = targetDistance
        self.targetTime = targetTime
        self.startDate = startDate
        self.endDate = endDate
        self.weeklySchedule = weeklySchedule
        self.isActive = isActive
        self.createdAt = createdAt
    }
}

enum GoalType: String, Codable, CaseIterable {
    case first5K = "FIRST_5K"
    case improve5K = "IMPROVE_5K"
    case first10K = "FIRST_10K"
    case improve10K = "IMPROVE_10K"
    case halfMarathon = "HALF_MARATHON"
    case marathon = "MARATHON"
    case generalFitness = "GENERAL_FITNESS"
    case weightLoss = "WEIGHT_LOSS"
    case custom = "CUSTOM"
}

struct ScheduledWorkout: Codable, Hashable, Identifiable {
    var id: String
    var dayOfWeek: Int
    var weekNumber: Int
    var workoutType: WorkoutType
    var targetDistanceMeters: Double?
    var targetDurationMinutes: Int?
    var targetPaceSecondsPerKm: Double?
    var targetPaceMinSecondsPerKm: Double?
    var targetPaceMaxSecondsPerKm: Double?
    var targetHeartRateZone: HeartRateZone?
    var targetHeartRateMin: Int?
    var targetHeartRateMax: Int?
    var intervals: [Interval]?
    var workoutDescription: String
    var isCompleted: Bool
    var completedRunId: UUID?

    init(
        id: String = UUID().uuidString,
        dayOfWeek: Int,
        weekNumber: Int,
        workoutType: WorkoutType,
        targetDistanceMeters: Double? = nil,
        targetDurationMinutes: Int? = nil,
        targetPaceSecondsPerKm: Double? = nil,
        targetPaceMinSecondsPerKm: Double? = nil,
        targetPaceMaxSecondsPerKm: Double? = nil,
        targetHeartRateZone: HeartRateZone? = nil,
        targetHeartRateMin: Int? = nil,
        targetHeartRateMax: Int? = nil,
        intervals: [Interval]? = nil,
        workoutDescription: String,
        isCompleted: Bool = false,
        completedRunId: UUID? = nil
    ) {
        self.id = id
        self.dayOfWeek = dayOfWeek
        self.weekNumber = weekNumber
        self.workoutType = workoutType
        self.targetDistanceMeters = targetDistanceMeters
        self.targetDurationMinutes = targetDurationMinutes
        self.targetPaceSecondsPerKm = targetPaceSecondsPerKm
        self.targetPaceMinSecondsPerKm = targetPaceMinSecondsPerKm
        self.targetPaceMaxSecondsPerKm = targetPaceMaxSecondsPerKm
        self.targetHeartRateZone = targetHeartRateZone
        self.targetHeartRateMin = targetHeartRateMin
        self.targetHeartRateMax = targetHeartRateMax
        self.intervals = intervals
        self.workoutDescription = workoutDescription
        self.isCompleted = isCompleted
        self.completedRunId = completedRunId
    }
}

enum WorkoutType: String, Codable, CaseIterable {
    // Basic runs
    case easyRun = "EASY_RUN"
    case longRun = "LONG_RUN"
    case recoveryRun = "RECOVERY_RUN"

    // Speed work
    case tempoRun = "TEMPO_RUN"
    case intervalTraining = "INTERVAL_TRAINING"
    case fartlek = "FARTLEK"
    case racePace = "RACE_PACE"

    // Strength & hills
    case hillRepeats = "HILL_REPEATS"
    case hillSprints = "HILL_SPRINTS"
    case stairWorkout = "STAIR_WORKOUT"

    // Specialized
    case progressionRun = "PROGRESSION_RUN"
    case negativeSplit = "NEGATIVE_SPLIT"
    case thresholdRun = "THRESHOLD_RUN"
    case vo2MaxIntervals = "VO2_MAX_INTERVALS"
    case yasso800s = "YASSO_800S"
    case mileRepeats = "MILE_REPEATS"
    case ladderWorkout = "LADDER_WORKOUT"
    case pyramidWorkout = "PYRAMID_WORKOUT"

    // Race simulation
    case raceSimulation = "RACE_SIMULATION"
    case timeTrial = "TIME_TRIAL"
    case parkrunPrep = "PARKRUN_PREP"

    // Recovery & maintenance
    case shakeOutRun = "SHAKE_OUT_RUN"
    case baseBuilding = "BASE_BUILDING"
    case aerobicMaintenance = "AEROBIC_MAINTENANCE"

    // Cross-training
    case crossTraining = "CROSS_TRAINING"
    case cycling = "CYCLING"
    case swimming = "SWIMMING"
    case strengthTraining = "STRENGTH_TRAINING"
    case yogaStretching = "YOGA_STRETCHING"

    // Rest
    case restDay = "REST_DAY"
    case activeRecovery = "ACTIVE_RECOVERY"

    // Custom
    case custom = "CUSTOM"
}

struct Interval: Codable, Hashable {
    var type: IntervalType
    var durationSeconds: Int?
    var distanceMeters: Double?
    var targetPaceSecondsPerKm: Double?
    var targetPaceMinSecondsPerKm: Double?
    var targetPaceMaxSecondsPerKm: Double?
    var targetHeartRateZone: HeartRateZone?
    var targetHeartRateMin: Int?
    var targetHeartRateMax: Int?
    var repetitions: Int

    init(
        type: IntervalType,
        durationSeconds: Int? = nil,
        distanceMeters: Double? = nil,
        targetPaceSecondsPerKm: Double? = nil,
        targetPaceMinSecondsPerKm: Double? = nil,
        targetPaceMaxSecondsPerKm: Double? = nil,
        targetHeartRateZone: HeartRateZone? = nil,
        targetHeartRateMin: Int? = nil,
        targetHeartRateMax: Int? = nil,
        repetitions: Int = 1
    ) {
        self.type = type
        self.durationSeconds = durationSeconds
        self.distanceMeters = distanceMeters
        self.targetPaceSecondsPerKm = targetPaceSecondsPerKm
        self.targetPaceMinSecondsPerKm = targetPaceMinSecondsPerKm
        self.targetPaceMaxSecondsPerKm = targetPaceMaxSecondsPerKm
        self.targetHeartRateZone = targetHeartRateZone
        self.targetHeartRateMin = targetHeartRateMin
        self.targetHeartRateMax = targetHeartRateMax
        self.repetitions = repetitions
    }
}

enum IntervalType: String, Codable, CaseIterable {
    case warmup = "WARMUP"
    case work = "WORK"
    case recovery = "RECOVERY"
    case cooldown = "COOLDOWN"
}

enum HeartRateZone: String, Codable, CaseIterable {
    case zone1 = "ZONE_1"
    case zone2 = "ZONE_2"
    case zone3 = "ZONE_3"
    case zone4 = "ZONE_4"
    case zone5 = "ZONE_5"

    var displayName: String {
        switch self {
        case .zone1: return "Recovery"
        case .zone2: return "Aerobic Base"
        case .zone3: return "Aerobic Capacity"
        case .zone4: return "Threshold"
        case .zone5: return "VO2 Max"
        }
    }

    var minPercent: Int {
        switch self {
        case .zone1: return 50
        case .zone2: return 60
        case .zone3: return 70
        case .zone4: return 80
        case .zone5: return 90
        }
    }

    var maxPercent: Int {
        switch self {
        case .zone1: return 60
        case .zone2: return 70
        case .zone3: return 80
        case .zone4: return 90
        case .zone5: return 100
        }
    }
}

struct HeartRateZoneTime: Codable, Hashable {
    var zone: HeartRateZone
    var durationMillis: Int64
    var percentOfTotal: Double

    var durationFormatted: String {
        let totalSeconds = durationMillis / 1000
        let minutes = totalSeconds / 60
        let seconds = totalSeconds % 60
        return String(format: "%d:%02d", minutes, seconds)
    }

    init(zone: HeartRateZone, durationMillis: Int64, percentOfTotal: Double = 0.0) {
        self.zone = zone
        self.durationMillis = durationMillis
        self.percentOfTotal = percentOfTotal
    }
}
